Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -23,586 +23,642 @@
-    //         -> !v1.has(c) -> requestSatisfyUtil(c, v1) -> v1.add(c) -> requestInstantiate1(c)
-    //     Background Executions = [ requestInstantiate1(r2), requestInstantiate1(a), requestInstantiate1(c) ]
-    //
-    // Step 4: requestInstantiate1(a) is done then go for it's dependencies [ b ]
-    //         -> v1.has(b) -> requestInstantiate2(b)
-    //         -> Since requestInstantiate1(b) cached in Step 3 and b is the only dependency a has, requestSatisfyUtil(a, v1) is fulfilled and cached to Entry(a).satisfy.
-    //     Background Executions = [ requestInstantiate1(r2), requestInstantiate1(c) ]
-    //
-    // Step 5: requestInstantiate1(r2) is done then go for it's dependencies [ a ]
-    //         -> !v2.has(a) -> requestSatisfyUtil(a, v2)
-    //         -> Since requestSatisfyUtil(a) is cached in Step 4 and a is the only dependency r2 has, requestSatisfyUtil(r2, v2) is fulfilled and cached.
-    //         -> linkAndEvaluateModule(r2) -> link(r2) crashes since c is not instantiated.
-    //     Background Executions = [ requestInstantiate1(c) ]
-    // 
+    if (entry.status === @ModuleStatusLinking || entry.status === @ModuleStatusLinked || entry.status === @ModuleStatusEvaluatingAsync || entry.status === @ModuleStatusEvaluated)
+        return index;
+
+    @assert(entry.status === @ModuleStatusUnlinked);
+    entry.status = @ModuleStatusLinking;
+    entry.dfsIndex = index;
+    entry.dfsAncestorIndex = index;
+    index++;
+    @arrayPush(stack, entry);
+
+    for (var i = 0, { length } = entry.requestedModules; i < length; i++) {
+        var specifier = entry.requestedModules[i];
+        var loadedEntry = entry.loadedModules.@get(specifier); // This implements step 2 of GetImportedModule
+        @assert(!!loadedEntry);
+        index = this.innerModuleLinking(loadedEntry, stack, index, fetcher);
+        @assert(loadedEntry.status === @ModuleStatusLinking || loadedEntry.status === @ModuleStatusLinked || loadedEntry.status === @ModuleStatusEvaluatingAsync || loadedEntry.status === @ModuleStatusEvaluated);
+        @assert((loadedEntry.status === @ModuleStatusLinking) === stack.@includes(loadedEntry));
+
+        if (loadedEntry.status === @ModuleStatusLinking)
+            entry.dfsAncestorIndex = @min(entry.dfsAncestorIndex, loadedEntry.dfsAncestorIndex);
+    }
 
-    "use strict";
-    
-    var satisfyingEntries = new @Set;
-    return this.requestSatisfyUtil(entry, parameters, fetcher, visited, satisfyingEntries).then((entry) => {
-        satisfyingEntries.@forEach((satisfyingEntry) => {
-            @cacheSatisfy(satisfyingEntry);
-            satisfyingEntry.isSatisfied = true;
-        });
-        return entry;
-    });
+    if (!entry.moduleRecord && entry.errorToRethrow)
+        throw entry.errorToRethrow;
+
+    entry.hasTLA = this.linkModuleRecord(entry.moduleRecord, fetcher);
+
+    @assert(stack.@filter(e => e === entry).length === 1);
```

Comment:
Do not use `filter`.

Reply:
Replaced with a simple `@countOf` helper.